src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageParser.cs

```diff
@@ -10,36 +10,47 @@ public static class BinaryMessageParser
     {
         public static bool TryParseMessage(ref ReadOnlyBuffer<byte> buffer, out ReadOnlyBuffer<byte> payload)
         {
-            long length = 0;
-            payload = default(ReadOnlyBuffer<byte>);
+            payload = default;
 
-            if (buffer.Length < sizeof(long))
+            if (buffer.IsEmpty)
             {
                 return false;
             }
 
-            // Read the length
-            length = buffer.Span.Slice(0, sizeof(long)).ReadBigEndian<long>();
+            var length = 0U;
+            var numBytes = 0;
 
-            if (length > Int32.MaxValue)
+            byte byteRead;
+            do
             {
-                throw new FormatException("Messages over 2GB in size are not supported");
+                byteRead = buffer.Span.Slice(numBytes, sizeof(byte)).Read<byte>();
+                length = length | (((uint)(byteRead & 0x7f)) << (numBytes * 7));
```

Comment:
Since we're bit twiddling, you can multiply by 7 like so `((n<<3) - n)` instead of using *.

Reply:
Yes

Reply:
And if we know the range of `numBytes` (between 0 and 5) then you can use a lookup table for possibly better speed. These need to be tested before actually done though...